demos/tutorials/RandomVariables.py

```diff
@@ -0,0 +1,384 @@
+# random variables (e.g. the covariance of y is lazily evaluated as 1/s).
+
+d = Gamma(1, 1e-4).rv
+s = Gamma(1, 1e-4).rv
+x = LMRF(0, 1/d, geometry=A.domain_geometry).rv
+y = Gaussian(A @ x, 1/s).rv
+
+# %%
+######################################################################
+# For illustrative purposes let us put this into a Bayesian Problem
+# object:
+# 
+
+BP = BayesianProblem(d, s, x, y)
+print(BP)
+
+# %%
+######################################################################
+# Providing data to the Bayesian problem creates a posterior as shown.
+#
+
+BP.set_data(y=y_obs)
+print(BP)
+
+# %%
+######################################################################
+# Given the Bayesian problem posterior we can now sample from
+# the posterior distribution using MCMC as follows:
+#
+
+samples = BP.sample_posterior(1000)
+
+# %%
+######################################################################
+# We can now plot a 95 credibility interval of the samples for x
+# from the posterior distribution as follows:
+#
+
+samples["x"].plot_ci(exact=info.exactSolution)
+
+# %%
+# Revisiting the random variable methods and attributes
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# In this section we revisit the methods and attributes of random variables
+# introduced earlier in this tutorial.
+# 
+# For this purpose let us define a new random 3-dimensional
+# variable :math:`x` as follows:
+
+x = Gaussian(0, 1, geometry=3).rv
+
+# %%
+######################################################################
+# Probability density evaluation
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# Probability density evaluation works by evaluating the probability
+# density function of the random variable at a given point. This is done
+# using the ``logd`` method. **Note**. Only simple (non-transformed)
+# random variables support this method at this point in time.
+#
+
+x.logd([1, 2, 3])
+
+# %%
+######################################################################
+# Gradient evaluation
+# ^^^^^^^^^^^^^^^^^^^
+#
+# Gradient evaluation works by evaluating the gradient of the log
+# probability density function of the random variable at a given point.
+#
+
+x.gradient([1, 2, 3])
+
+
+# %%
+######################################################################
+# Dimension and geometry
+# ^^^^^^^^^^^^^^^^^^^^^^
+#
+# The dimension of a random variable can be accessed using the ``dim``
+# attribute. The geometry of a random variable can be accessed using the
+# ``geometry`` attribute.
+#
+
+print(x.dim)
+print(x.geometry)
+
+# %%
+######################################################################
+# Conditioning
+# ^^^^^^^^^^^^
+#
+# Conditioning works by conditioning the underlying distribution of the
+# random variable on the given input. This is done using the ``condition``
+# method. For more details see documentation of conditioning on distributions.
+
+z = Gaussian(0, lambda s: s).rv
+
+print(z)
+print(z.condition(s=10))
+
```

Comment:
I tried creating the conditioned distribution first, then condition it, then generate the rv, and get an error:
```
Z = Gaussian(0, lambda s: s)
zz = Z(s=3)
zz.rv
```
this is the error
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/.local/lib/python3.9/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/.local/lib/python3.9/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    392                         if cls is not object \
    393                                 and callable(cls.__dict__.get('__repr__')):
--> 394                             return _repr_pprint(obj, self, cycle)
    395 
    396             return _default_pprint(obj, self, cycle)

~/.local/lib/python3.9/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    698     """A pprint that just redirects to the normal repr function."""
    699     # Find newlines and replace them with p.break_()
--> 700     output = repr(obj)
    701     lines = output.splitlines()
    702     with p.group():

~/Work/CUQI/dev/CUQIpy/cuqi/randomvariable/_randomvariable.py in __repr__(self)
    217     def __repr__(self):
    218         # Create strings for parameter name ~ distribution pairs
--> 219         parameter_strings = [f"{name} ~ {distribution}" for name, distribution in zip(self.parameter_names, self.distributions)]
    220         # Join strings with newlines
    221         parameter_strings = "\n".join(parameter_strings)

~/Work/CUQI/dev/CUQIpy/cuqi/randomvariable/_randomvariable.py in parameter_names(self)
    195     def parameter_names(self) -> str:
    196         """ Name of the parameter that the random variable can be evaluated at. """
--> 197         return [distribution.par_name for distribution in self.distributions]
    198 
    199     @property

~/Work/CUQI/dev/CUQIpy/cuqi/randomvariable/_randomvariable.py in distributions(self)
    179         # Inject name into distributions
    180         if len(self._distributions) == 1:
--> 181             self._set_dist_name_if_not_set(next(iter(self._distributions)))
    182         return self._distributions
    183 

~/Work/CUQI/dev/CUQIpy/cuqi/randomvariable/_randomvariable.py in _set_dist_name_if_not_set(self, distribution)
    184     def _set_dist_name_if_not_set(self, distribution):
    185         if not hasattr(distribution, '_par_name') or distribution._par_name is None: #TODO. Change to "par_name" once dist does not auto infer
--> 186             distribution.par_name = self.name
    187 
    188     def get_conditioning_variables(self):

~/Work/CUQI/dev/CUQIpy/cuqi/density/_density.py in par_name(self, name)
     45     def par_name(self, name):
     46         if self._is_copy:
---> 47             raise ValueError("Cannot set name of conditioned density. Only the original density can have its name set.")
     48         self._par_name = name
     49 

ValueError: Cannot set name of conditioned density. Only the original density can have its name set.
```

Reply:
I updated such that this can be done now if the original density have not yet a parameter name set. One has to do

```python
Z = Gaussian(0, lambda s: s)
zz = Z(s=3)
z = zz.rv # Name and parameter name becomes z here.
```